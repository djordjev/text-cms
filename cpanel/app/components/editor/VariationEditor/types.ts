export enum Action {
  Bold = 'bold',
  Italic = 'italic',
  Underline = 'underline',
  Strikethrough = 'strikethrough',
  Color = 'color'
}
