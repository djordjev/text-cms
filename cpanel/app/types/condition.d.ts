export type ConditionOperator = '=' | '>' | '>=' | '<' | '<=' | 'in';

export type ConditionDescriptor = [string, ConditionOperator, string];

export type ConditionAndChain = ConditionDescriptor[];

export type ConditionGroup = ConditionAndChain[];
