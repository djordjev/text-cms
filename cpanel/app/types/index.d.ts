export type CreateParam = 'new-file' | 'new-folder';

export type FormValues = { [p: string]: FormDataEntryValue };

export type FileVariation = {
  condition?: string;
  id: string;
  name: string;
  text: string;
};

export type Payload<T> = {
  data?: T;
  error?: string;
  errors?: Record<string, string>;
};
