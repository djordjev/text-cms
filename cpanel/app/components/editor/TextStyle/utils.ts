import { BaseEditor, Editor, Element } from 'slate';
import { ReactEditor } from 'slate-react';

import { CustomElement } from '~/components/editor/VariationEditor/types';

const getSelectedStyle = (editor: BaseEditor & ReactEditor) => {
  const { selection } = editor;

  if (!selection) return;

  const [match] = Array.from(
    Editor.nodes<CustomElement>(editor, {
      at: selection,
      match: (n) =>
        !Editor.isEditor(n) &&
        Element.isElement(n) &&
        (n.type === 'paragraph' || n.type === 'heading')
    })
  );

  const [nodeType] = match;

  return nodeType;
};

export { getSelectedStyle };
