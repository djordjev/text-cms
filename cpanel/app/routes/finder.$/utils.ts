import { CreateParam } from '~/types';

const getCreateOpenModal = (params: URLSearchParams): CreateParam | null => {
  const param = params.get('action');

  if (param === 'new-file' || param === 'new-folder') return param;

  return null;
};

export { getCreateOpenModal };
