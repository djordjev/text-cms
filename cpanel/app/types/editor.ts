import { BaseEditor } from 'slate';
import { ReactEditor } from 'slate-react';

export enum Action {
  Bold = 'bold',
  Click = 'click',
  Color = 'color',
  Italic = 'italic',
  Strikethrough = 'strikethrough',
  Underline = 'underline'
}

type StrAct = Action.Color;
type BoolAct = Exclude<Action, StrAct>;
type ClickAct = Action.Click;

export type Alignment = 'left' | 'center' | 'right';

export type ElementCommon = {
  align?: Alignment;
};

export type ParagraphElement = {
  type: 'paragraph';
  children: CustomText[];
} & ElementCommon;

export type HeadingElement = {
  type: 'heading';
  level: number;
  children: CustomText[];
} & ElementCommon;

export type CustomElement = ParagraphElement | HeadingElement;

export type ClickActionPayload = {
  action?: string;
  href?: string;
  type: 'link' | 'primary' | 'secondary';
};

type StringActions = { [key in StrAct]?: string };
type BoolActions = { [key in BoolAct]?: boolean };
type ClickActions = { [key in ClickAct]?: ClickActionPayload };

export type AnyAction = string | boolean | ClickActionPayload;

export type CustomText = { text: string } & BoolActions &
  StringActions &
  ClickActions;

declare module 'slate' {
  interface CustomTypes {
    Editor: BaseEditor & ReactEditor;
    Element: CustomElement;
    Text: CustomText;
  }
}
