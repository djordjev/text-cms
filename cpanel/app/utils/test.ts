import { Descendant } from 'slate';

export const TEST_EDITOR: Descendant[] = [
  {
    type: 'heading',
    children: [
      {
        text: 'HEADING'
      }
    ],
    level: 3,
    align: 'center'
  },
  {
    type: 'paragraph',
    children: [
      {
        text: 'A '
      },
      {
        text: 'line',
        italic: true
      },
      {
        text: ' of '
      },
      {
        text: 'text',
        color: '#e41111',
        bold: true
      },
      {
        text: ' in a '
      },
      {
        text: 'paragraph',
        bold: true
      },
      {
        text: '.'
      }
    ]
  }
];

export const buildRequest = (data: Record<string, string>, method = 'POST') => {
  const form = new URLSearchParams(data);

  return new Request('https://www.test.local', { body: form, method });
};
