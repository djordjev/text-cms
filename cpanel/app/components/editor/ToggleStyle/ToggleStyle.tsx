import {
  IconBold,
  IconItalic,
  IconStrikethrough,
  IconUnderline,
  TablerIconsProps
} from '@tabler/icons-react';
import classnames from 'classnames';
import { FC, MouseEvent } from 'react';
import { Editor } from 'slate';
import { useSlate } from 'slate-react';

import { Action } from '~/types/editor';

export interface ToggleStyleProps {
  action: Action;
  className?: string;
  onToggle: (isSelected: boolean) => void;
}

const ToggleStyle: FC<ToggleStyleProps> = (props) => {
  const { action, className, onToggle } = props;

  // Hooks
  const editor = useSlate();

  // Setup
  const marks = Editor.marks(editor);
  const active = !!marks?.[action];

  const iconProps: TablerIconsProps = { height: 20, width: 20 };

  // Styles
  const classes = classnames('u-p-1xs', className, {
    'u-bg-accent u-rounded-md': active
  });

  // Handlers
  const onClick = (e: MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    onToggle(!active);
  };

  // Markup
  const renderIcon = () => {
    if (action === Action.Bold) return <IconBold {...iconProps} />;
    if (action === Action.Italic) return <IconItalic {...iconProps} />;
    if (action === Action.Underline) return <IconUnderline {...iconProps} />;
    if (action === Action.Strikethrough) {
      return <IconStrikethrough {...iconProps} />;
    }

    return null;
  };

  return (
    <button className={classes} onMouseDown={onClick} type="button">
      {renderIcon()}
    </button>
  );
};

export { ToggleStyle };
