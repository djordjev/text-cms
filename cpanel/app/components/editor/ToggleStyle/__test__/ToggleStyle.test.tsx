import { createRemixStub } from '@remix-run/testing';
import { fireEvent, render, screen } from '@testing-library/react';
import { createEditor } from 'slate';
import { Slate, withReact } from 'slate-react';

import { Action } from '~/types/editor';
import { TEST_EDITOR } from '~/utils/test';

import { ToggleStyle, ToggleStyleProps } from '../ToggleStyle';

test('ToggleStyle', () => {
  const onToggle = vi.fn();
  const editor = withReact(createEditor());

  const props: ToggleStyleProps = {
    action: Action.Bold,
    onToggle
  };

  const createComponent = () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: () => (
          <Slate editor={editor} initialValue={TEST_EDITOR}>
            <ToggleStyle {...props} />{' '}
          </Slate>
        )
      }
    ]);

    return <RemixStub />;
  };

  // Bold
  const { container, rerender } = render(createComponent());
  let icon = screen.getByRole('img');
  expect(icon).toHaveClass('tabler-icon-bold');
  expect(container.childNodes).toMatchSnapshot('bold');

  // Italic
  props.action = Action.Italic;
  rerender(createComponent());
  icon = screen.getByRole('img');
  expect(icon).toHaveClass('tabler-icon-italic');
  expect(container.childNodes).toMatchSnapshot('italic');

  // Underline
  props.action = Action.Underline;
  rerender(createComponent());
  icon = screen.getByRole('img');
  expect(icon).toHaveClass('tabler-icon-underline');
  expect(container.childNodes).toMatchSnapshot('underline');

  // Strikethrough
  props.action = Action.Strikethrough;
  rerender(createComponent());
  icon = screen.getByRole('img');
  expect(icon).toHaveClass('tabler-icon-strikethrough');
  expect(container.childNodes).toMatchSnapshot('strikethrough');

  // Handles click
  fireEvent.mouseDown(screen.getByRole('button'));
  expect(onToggle).toHaveBeenCalledOnce();
  expect(onToggle).toHaveBeenCalledWith(true);
});
