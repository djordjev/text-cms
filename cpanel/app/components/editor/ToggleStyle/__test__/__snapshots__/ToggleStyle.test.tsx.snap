// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleStyle > bold 1`] = `
NodeList [
  <button
    class="u-p-1xs"
    type="button"
  >
    <svg
      class="tabler-icon tabler-icon-bold"
      fill="none"
      height="20"
      role="img"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5h6a3.5 3.5 0 0 1 0 7h-6z"
      />
      <path
        d="M13 12h1a3.5 3.5 0 0 1 0 7h-7v-7"
      />
    </svg>
  </button>,
   ,
]
`;

exports[`ToggleStyle > italic 1`] = `
NodeList [
  <button
    class="u-p-1xs"
    type="button"
  >
    <svg
      class="tabler-icon tabler-icon-italic"
      fill="none"
      height="20"
      role="img"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11 5l6 0"
      />
      <path
        d="M7 19l6 0"
      />
      <path
        d="M14 5l-4 14"
      />
    </svg>
  </button>,
   ,
]
`;

exports[`ToggleStyle > strikethrough 1`] = `
NodeList [
  <button
    class="u-p-1xs"
    type="button"
  >
    <svg
      class="tabler-icon tabler-icon-strikethrough"
      fill="none"
      height="20"
      role="img"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M5 12l14 0"
      />
      <path
        d="M16 6.5a4 2 0 0 0 -4 -1.5h-1a3.5 3.5 0 0 0 0 7h2a3.5 3.5 0 0 1 0 7h-1.5a4 2 0 0 1 -4 -1.5"
      />
    </svg>
  </button>,
   ,
]
`;

exports[`ToggleStyle > underline 1`] = `
NodeList [
  <button
    class="u-p-1xs"
    type="button"
  >
    <svg
      class="tabler-icon tabler-icon-underline"
      fill="none"
      height="20"
      role="img"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5v5a5 5 0 0 0 10 0v-5"
      />
      <path
        d="M5 19h14"
      />
    </svg>
  </button>,
   ,
]
`;
